rfcs/2020-04-06-1858-automatically-adjust-request-limits.md

```diff
@@ -0,0 +1,257 @@
+        let future = self.inner.call(request);
+        ...
+        emit!(ConcurrencyLimit { concurrency: self.limit.maximum() });
+        emit!(ConcurrencyActual { concurrency: self.limit.used() });
+        ResponseFuture::new(future, self.limit.semaphore.clone(), Instant::now())
+    }
+}
+
+impl Future for ResponseFuture {
+    fn poll(&mut self, cx: &mut Context) -> Poll<Self::Output> {
+        match self.inner.poll() {
+            Pending => Pending,
+            Ready(output) => {
+                let now = Instant::now();
+                let rtt = now.duration_since(self.start_time);
+                emit!(RTTMeasurement { rtt: rtt.as_millis() });
+
+                if now >= controller.next_update {
+                    let mut controller = self.controller.lock();
+                    if rtt > controller.rtt + controller.threshold {
+                        // The `+ 1` prevents this from going to zero
+                        controller.concurrency = (controller.concurrency + 1) / 2;
+                    }
+                    else if controller.concurrency < controller.in_flight_limit
+                        && rtt <= controller.rtt {
+                        controller.concurrency += 1;
+                    }
+                    controller.next_update = now + controller.average_rtt
+                }
+                controller.average_rtt.update(rtt);
+
+                Ready(output)
+            }
+        }
+    }
+}
+```
+
+### Observed Behavior
+
+This algorithm should have the following responses to service
+conditions:
+
+* Under normal use, the RTT will stay relatively constant or increase
+  slightly in proportion with the concurrency. This should allow the
+  concurrency to increase slowly to the configured maximum, increasing
+  the delivery rate (assuming no limit is reached).
+
+* If a remote service suddenly becomes unresponsive, with sustained
+  timeouts, Vector will rapidly reduce the request concurrency down to
+  the minimum of one.
+
+* If a remote service gradually increase its response time, Vector will
+  gracefully reduce its request concurrency, with it going down to the
+  minimum if the response time continues to increase.
+
+* If a remote service has a hard rate limit, expressed with either HTTP
+  response 429 or timeouts for example, lower than what Vector has ready
+  to deliver to it, Vector's concurrency should hover around `rate_limit
+  / RTT`, peeking over and then briefly dropping down when queries are
+  limited. This will keep the delivery rate close to the discovered rate
+  limit.
+
+### Observability
+
+Vector operators need to be able to observe the behavior of this
+algorithm to ensure that it is operating as desired. To this end, the
+mechanism will expose the following data:
+
+* a rate-limited event logged every time a remote service explicitly
+  limits the request rate:
+  
+  `Requests throttled by remote service, concurrency reduced.`
```

Comment:
I don't think this is accurate. Vector should receive a `429` response and that'll be logged like any other error response, right? What we want, instead, is to log when Vector _internally_ rate limits. That is, your algorithm has adjusted the `in_flight_limit` to 10 and an 11th request gets rate limited. When that happens we should replace this log:

```
TRACE tower_limit::rate::service: rate limit exceeded, disabling service
```

With this log

```
WARN sink{name=out type=http} vector::sinks::http has been internally rate-limited, `requests.in_flight_limit` is set to 10 and this is request 11.
```

What do you think?

Reply:
This has been discussed in #1391, but I think it's very important we communicate this properly to users, so it should be included in the scope of this RFC.

Reply:
I agree that would be a useful thing to report, and clears up some misunderstanding. It does sound, however, like something better as a metric than a log message. Even rate limited we likely want to have a counter on how often it happens.